subsys: ipc: add more tests on nrf5340bsim

dsseng opened the pull request:
Make sure IPC subsystem is easy to test on BabbleSim for better CI coverage.

- **samples: ipc: icmsg: set EP name**
- **samples: ipc: icmsg: test icbmsg on nrf5340bsim**
- ...

dsseng committed:
tests: subsys: ipc: ipc_sessions: fix wrong sizeof usage

Fix clangd warnings around use of sizeof on pointers.

No functional change.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
tests: subsys: ipc: ipc_sessions: decrement before sending

Once the primary CPU receives the block, it will immediately try to ask
the remote for blocks left to send.

On real hardware, this would typically work fine, since the IRQ
round-trip allows remote to finish the loop iteration.

However on nrf5340bsim the host answers immediately, receiving wrong
data and failing the test.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
ipc: icmsg: fix unbound for POSIX

Make sure unbound works on nrf5340bsim.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
tests: subsys: ipc: ipc_sessions: run on nrf5340bsim

Improve CI coverage.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
tests: subsys: ipc: ipc_sessions: add _no_unbound tests for bsim

Also cover these code paths.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented:
cc @aescolar regarding #94364

dsseng commented on tests/subsys/ipc/ipc_sessions/testcase.yaml:
This is not looking like it's working well, so consider the PR completed unless I add that test before merge

dsseng commented on tests/subsys/ipc/ipc_sessions/testcase.yaml:
```diff
@@ -8,28 +8,46 @@ common:
       - nrf5340dk/nrf5340/cpuapp
-  sample.ipc.ipc_sessions.nrf54h20dk_cpuapp_cpurad:
+      - nrf5340bsim/nrf5340/cpuapp
+  ipc.ipc_sessions.nrf5340bsim_cpuapp_no_unbound_cpunet:
```
These should also work with hardware, but as I have not tested that I only added `nrf5340bsim` to the integration platforms

dsseng commented on tests/subsys/ipc/ipc_sessions/testcase.yaml:
```diff
@@ -8,28 +8,46 @@ common:
 
 tests:
-  sample.ipc.ipc_sessions.nrf5340dk:
+  ipc.ipc_sessions.nrf5340:
```
Thanks, will apply with the next batch of changes

kartben commented on include/zephyr/ipc/pbuf.h:
```diff
@@ -269,6 +269,20 @@ void pbuf_handshake_write(struct pbuf *pb, uint32_t value);
  */
 int pbuf_get_initial_buf(struct pbuf *pb, volatile char **buf, uint16_t *len);
 
+#if defined(CONFIG_ARCH_POSIX)
```
don't guard the declaration in the header file, it's unnecessary

dsseng replied:
This makes sense, thank you. It's anyway documented as an internal matter, so users don't call it

kartben commented on samples/subsys/ipc/ipc_service/icmsg/sample.yaml:
```diff
@@ -89,6 +89,34 @@ tests:
         - "host: Received"
         - "host: IPC-service HOST demo ended"
 
+  sample.ipc.icbmsg.nrf5340bsim:
+    platform_allow:
+      - nrf5340bsim/nrf5340/cpuapp
+    integration_platforms:
+      - nrf5340bsim/nrf5340/cpuapp
+    tags: ipc
+    extra_args:
+      - icmsg_DTC_OVERLAY_FILE="boards/nrf5340bsim_nrf5340_cpuapp_icbmsg.overlay"
+      - remote_DTC_OVERLAY_FILE="boards/nrf5340bsim_nrf5340_cpunet_icbmsg.overlay"
+    sysbuild: true
+    harness: console
+    harness_config:
+      type: multi_line
+      ordered: false
+      regex:
+        - "host: IPC-service HOST demo started"
+        - "host: Ep bounded"
+        - "host: Perform sends for"
+        - "host: Sent"
+        - "host: Received"
+        - "host: IPC-service HOST demo ended"
+        - "remote: IPC-service REMOTE demo started"
+        - "remote: Ep bounded"
+        - "remote: Perform sends for"
+        - "remote: Sent"
+        - "remote: Received"
+        - "remote: IPC-service REMOTE demo ended"
```
please don't use sample as test

dsseng replied:
Yes, as previously discussed, this should be migrated to a new test case, covering icbmsg and some other features current test cases do not cover

Also, nrf5340dk is the recommended target for this sample, so perhaps testing on it in a simulated environment can fit the purpose of testing the sample functionality

dsseng commented:
@doki-nordic @JarmouniA - could you please take a (another) look? Thank you!

dsseng commented:
@doki-nordic @kartben @JarmouniA could you please review this PR since I have addressed all the actionable comments (i.e. not questions to others)? Thank you!

aescolar commented on tests/subsys/ipc/ipc_sessions/remote/src/remote.c:
```diff
@@ -432,6 +432,7 @@ int main(void)
 				}
 				do {
 					ret = ipc_service_send(ep_cfg.priv, cmd_data, cmd_size);
+					k_busy_wait(1);
```
@dsseng if this was added to allow time to pass in the posix arch, you can use instead `Z_SPIN_DELAY(1)`

dsseng committed:
tests: subsys: ipc: break busy loops

Busy loops without any way to yield to the kernel are not recommended,
and here it causes a multicore BabbleSim simulation to stall.

Co-authored-by: Alberto Escolar Piedras <alberto.escolar.piedras@nordicsemi.no>
Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>
